File: C04_InnerTube/InnerTubeVB/SharedUtilities/ExtensionMethods.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq
Imports System.Text

Namespace SharedUtilities
    Public Module ExtensionMethods

        <System.Runtime.CompilerServices.Extension()> _
        Public Function ToCollection(Of T)(ByVal source As IEnumerable(Of T)) As Collection(Of T)
            Dim result As Collection(Of T) = New Collection(Of T)()

            For Each item In source
                result.Add(item)
            Next item
            Return result
        End Function

        <System.Runtime.CompilerServices.Extension()> _
        Public Function ToObservableCollection(Of T)(ByVal source As IEnumerable(Of T)) As ObservableCollection(Of T)
            Dim result As ObservableCollection(Of T) = New ObservableCollection(Of T)()

            For Each item In source
                result.Add(item)
            Next item
            Return result
        End Function

    End Module
End Namespace

File: C04_InnerTube/InnerTubeCS/SharedUtilities/ExtensionMethods.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace SharedUtilities
{
    public static class ExtensionMethods
    {
        public static Collection<T> ToCollection<T>(this IEnumerable<T> source)
        {
            Collection<T> result = new Collection<T>();

            foreach (var item in source)
            {
                result.Add(item);
            }
            return result; 
        }

        public static ObservableCollection<T> ToObservableCollection<T>(this IEnumerable<T> source)
        {
            ObservableCollection<T> result = new ObservableCollection<T>();

            foreach (var item in source)
            {
                result.Add(item);
            }
            return result;
        }

    }
}
